matchbook-sdk-rest/src/main/java/com/matchbook/sdk/rest/PartiallyFailableResponseCallback.java

```diff
@@ -0,0 +1,89 @@
+package com.matchbook.sdk.rest;
+
+import com.matchbook.sdk.core.StreamObserver;
+import com.matchbook.sdk.core.exceptions.MatchbookSDKParsingException;
+import com.matchbook.sdk.rest.configs.Parser;
+import com.matchbook.sdk.rest.configs.Serializer;
+import com.matchbook.sdk.rest.dtos.FailableRestResponse;
+import com.matchbook.sdk.rest.dtos.PartiallyFailableResponse;
+import com.matchbook.sdk.rest.readers.Reader;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Objects;
+
+class PartiallyFailableResponseCallback<T extends FailableRestResponse, RESP extends PartiallyFailableResponse<T>>
+        extends RestResponseCallback<T, RESP> {
+
+    private static final int FAILABLE_HTTP_CODE = 400;
+
+    PartiallyFailableResponseCallback(StreamObserver<T> observer, Serializer serializer, Reader<T, RESP> reader) {
+        super(observer, serializer, reader);
+    }
+
+    @Override
+    public void onFailedResponse(InputStream responseInputStream, int responseCode) {
+        if (responseCode == FAILABLE_HTTP_CODE) {
+            onFailableResponse(responseInputStream, responseCode);
+        } else {
+            super.onFailedResponse(responseInputStream, responseCode);
+        }
+    }
+
+    private void onFailableResponse(InputStream responseInputStream, int responseCode) {
+        try (ByteArrayOutputStream responseBytes = copyBytes(responseInputStream);
+                InputStream copiedInputStream = copyInputStream(responseBytes)) {
```

Comment:
why do you copy the input stream?

Reply:
Unfortunately, the only way to know if the response contains a partially failed content or a full error message is to start parsing the bytes. If I parse the input stream, I'll consume it and make it unavailable. The only solution I know in order to be capable to parse the content twice is to copy the bytes is a swapping data structure, parse it once and only if necessary parse the copy another time.

Of course this way we can not take advantage of the stream parsers. However, we use this technique only for responses that might partially fail and only when the response HTTP code suggests it could be the case.

Reply:
yeah, i got you. Let's leave it like that at the moment, but we should improve this in a future